text/0044-heartbeat.md

```diff
@@ -0,0 +1,69 @@
+
+# Motivation
+
+As part of the [SDK onboarding improvements initiative](https://www.notion.so/sentry/SDK-Onboarding-Improvements-261a3d1deed94522bcff1361fc8bd756) (employees only), user tests have recently been done, where we collect data about the user experience when trying to integrate Sentry for the first time with their applications.
+
+Several discoveries were made, but one in particular that caught our attention is that many users, after following all the installation steps and sending the first error, were confused and frustrated if they did not see it in the Sentry's application.
+
+After refreshing the Sentry's application page a few times to see if they get an error and not getting it, they were left wondering if they missed something while configuring Sentry, which led them to open the onboarding documentation again.
+
+We also noticed that when they successfully sent an issue to Sentry, they were confused as they were taken to the issue stream page without receiving any feedback in the UI, which was initially awaiting for an event.
+
+# Background
+
+Historically there has never been supported use cases where Sentry has allowed SDKs would “call home” and check-in. Client Reports provided functionality which moved in that direction. Originally users could only see server-side “dropped” events, but client reports allowed us to gather numbers on how many events are “discarded” before ever leaving the SDK. It was originally designed as a solution to determine if a problem existed in any of the SDKs where we had a lot of discards due to technical issues. But what we have found is that sales and support find this information to be empowering, allowing them to see for the customer what their sample rate might be and tell them if they are really missing errors or transactions due to Sentry dropping or due to the SDK discarding for any number of outcome reasons. With this in mind, there are more use cases to be built on top of client reports. One of these is a ping endpoint functionality, similar to a health endpoint, so developers can check if their apps have the SDKs initialized correctly.
+
+# Supporting Data
+
+AJ:
+
+> I am looking at cohort where of all the orgs sending txn only 13% actually use perf
+
+I believe orgs sending txn rate is pretty good for new projects because most devs don’t look at the details in the code snippet and end up sending txn. However, the challenge is they don’t know about Perf or never go there
+
+# Options Considered
+
+## Heartbeat/Pulse endpoint
+
+Create an endpoint called heartbeat or pulse that would be used via a request whenever the user successfully launches their app.
+
+The request body would initially be quite simple, containing only the following:
+
+`{type: 'heartbeat'}`
+
+Sentry when collecting this information will be able to give feedback to the user through the user interface, such as that he can now send an error or perform other actions.
```

Comment:
We saw during the user testing sessions that we did, that after the user went through all the onboarding steps and sent an error, he would get frustrated if the error wasn't coming, making him wonder if he missed something during the configuration, for instance, if the DSN he used was correct. 

We just want to give the user the feedback that a connection was successfully established so he can discard, for instance the DSN,  from his troubleshooting list

Reply:
I get this, but we also can have here same problem. 

If we don't, in the implementation, introduce some kind of QoS for these heartbeats, they could be stuck somewhere in pipeline, can be delayed and won't give expected (fast) feedback. 

I'm talking about e.g: 
* network latency (which could be difficult to account for in some cases)
* or maybe the ingesting/processing queues are full with other types of messages and heartbeats won't be forwarded just right away

I'm not sure if this, QoS-related discussion, should be the part this RFC or later the part of the implementation. But I wanted just to mention it here so it won't get lost.

Reply:
great points

> For me, ability to send the errors / events / envelopes only shows that everything is working

i don't think we want to get to a place where the product can state "everything is configured correctly" that is a bold statement without knowledge of which frameworks and DB integrations they need, then if that SDK supports auto instrumentation for them etc.

But the product can at least start to say "hey you are definitely connected" as a step 1. Further iterations of this would include info about the options which have been configured in the init and set by default, sentry would know a bit more.all without the requirement of pushing them to send an error or txn

> QoS
yes I think we can say this out of scope for the time being :D